File: flow-sample-intents/src/androidTest/java/flow/sample/intents/IntentsSampleTest.java
package flow.sample.intents;

import android.content.Context;
import android.content.Intent;

import org.junit.Rule;
import org.junit.Test;

import java.util.List;

import androidx.test.rule.ActivityTestRule;
import flow.Flow;
import flow.History;

import static androidx.test.InstrumentationRegistry.getInstrumentation;
import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.pressBack;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static java.util.Arrays.asList;

/**
 * Demonstrates the use of Intents to drive espresso tests to specific screens.
 */
public class IntentsSampleTest {

  @Rule public ActivityTestRule rule = new ActivityTestRule<>(IntentsSingleInstanceSampleActivity.class);

  @Test public void singleInstanceActivity() {
    List<String> expected = asList("Able", "Baker", "Charlie");
    History h = History.emptyBuilder().pushAll(expected).build();

    Context context = getInstrumentation().getTargetContext().getApplicationContext();
    Intent intent = new Intent(context, IntentsSingleInstanceSampleActivity.class);
    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
    Flow.addHistory(intent, h, new StringParceler());
    context.startActivity(intent);

    onView(withText("Charlie")).perform(pressBack());
    onView(withText("Baker")).perform(pressBack());
    onView(withText("Able")).perform(click());
  }

  @Test public void standardActivity() {
    List<String> expected = asList("Higgledy", "Piggledy", "Pop");
    History h = History.emptyBuilder().pushAll(expected).build();

    Context context = getInstrumentation().getTargetContext().getApplicationContext();
    Intent intent = new Intent(context, IntentsStandardSampleActivity.class);
    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
    Flow.addHistory(intent, h, new StringParceler());
    context.startActivity(intent);

    onView(withText("Pop")).perform(pressBack());
    onView(withText("Piggledy")).perform(pressBack());
    onView(withText("Higgledy")).perform(click());
  }
}

File: flow-sample-intents/src/androidTest/java/flow/sample/intents/IntentsSampleTest.kt
package flow.sample.intents

import android.content.Intent
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.pressBack
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.platform.app.InstrumentationRegistry
import flow.Flow
import flow.historyOf
import org.junit.Test

/**
 * Demonstrates the use of Intents to drive espresso tests to specific screens.
 */
class IntentsSampleTest {
  @Test
  fun singleInstanceActivity() {
    val history = historyOf("Able", "Baker", "Charlie")

    val context = InstrumentationRegistry.getInstrumentation().targetContext
    val intent = Intent(context, IntentsSingleInstanceSampleActivity::class.java)
    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
    Flow.addHistory(intent, history, StringParceler())

    context.startActivity(intent)

    onView(withText("Charlie")).perform(pressBack())
    onView(withText("Baker")).perform(pressBack())
    onView(withText("Able")).perform(click())
  }

  @Test
  fun standardActivity() {
    val history = historyOf("Higgledy", "Piggledy", "Pop")

    val context = InstrumentationRegistry.getInstrumentation().targetContext
    val intent = Intent(context, IntentsStandardSampleActivity::class.java)
    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
    Flow.addHistory(intent, history, StringParceler())

    context.startActivity(intent)

    onView(withText("Pop")).perform(pressBack())
    onView(withText("Piggledy")).perform(pressBack())
    onView(withText("Higgledy")).perform(click())
  }
}